content/00.build/65.developer-reference/50.era-contracts/60.elliptic-curve-precompiles.md

```diff
@@ -0,0 +1,311 @@
+/// @param multiplier The multiplier in Montgomery form.
+/// @return ret The result of the Montgomery multiplication.
+function montgomeryMul(multiplicand, multiplier) -> ret {
+    let hi := getHighestHalfOfMultiplication(multiplicand, multiplier)
+    let lo := mul(multiplicand, multiplier)
+    ret := REDC(lo, hi)
+}
+
+```
+
+### Montgomery Inversion
+
+```solidity
+/// @notice Computes the Montgomery modular inverse skipping the Montgomery reduction step.
+/// @dev The Montgomery reduction step is kept because a modification in the binary extended Euclidean algorithm is used to compute the modular inverse.
+/// @dev See the function `binaryExtendedEuclideanAlgorithm` for further details.
+/// @param a The field element in Montgomery form to compute the modular inverse of.
+/// @return invmod The result of the Montgomery modular inverse (in Montgomery form).
+function montgomeryModularInverse(a) -> invmod {
+    invmod := binaryExtendedEuclideanAlgorithm(a)
+}
+```
+
+As said before, we use a modified version of the bEE algorithm that lets us “skip” the Montgomery reduction step.
+
+The regular algorithm would be $REDC((aR \mod N)^{−1}(R^3 \mod N))$ which involves a regular inversion plus a multiplication by a value that can be precomputed.
+
+## ECADD
+
+Precompile for computing elliptic curve point addition. The points are represented in affine form, given by a pair of coordinates $(x,y)$.
+
+Affine coordinates are the conventional way of expressing elliptic curve points, which use 2 coordinates. The math is concise and easy to follow.
+
+For a pair of constants $a$ and $b$, an elliptic curve is defined by the set of all points $(x,y)$
+that satisfy the equation $y^2=x^3+ax+b$, plus a special “point at infinity” named $O$.
+
+### Point Doubling
+
+To compute $2P$ (or $P+P$), there are three cases:
+
+- If $P = O$, then $2P = O$.
+- Else $P = (x, y)$
+  - If $y = 0$, then $2P = O$.
+  - Else $y≠0$, then
+
+        $$
+        \begin{gather*} \lambda = \frac{3x_{p}^{2} + a}{2y_{p}} \\ x_{r} = \lambda^{2} - 2x_{p} \\ y_{r} = \lambda(x_{p} - x_{r}) - y_{p}\end{gather*}
+        $$
+
+The complicated case involves approximately 6 multiplications, 4 additions/subtractions, and 1 division.
+There could also be 4 multiplications, 6 additions/subtractions, and 1 division,
+and if you want you could trade a multiplication with 2 more additions.
+
+### Point Addition
+
+To compute $P + Q$ where $P \neq Q$, there are four cases:
+
+- If $P = 0$ and $Q \neq 0$, then $P + Q = Q$.
+- If $Q = 0$ and $P \neq 0$, then $P + Q = P$.
+- Else $P = (x_{p},\ y_{p})$ and$Q = (x_{q},\ y_{q})$
+  - If $x_{p} = x_{q}$ (and necessarily $y_{p} \neq y_{q}$), then $P + Q = O$.
+  - Else $x_{p} \neq x_{q}$, then
+
+        $$
+        \begin{gather*} \lambda = \frac{y_{2} - y_{1}}{x_{2} - x_{1}} \\ x_{r} = \lambda^{2} - x_{p} - x_{q} \\ y_{r} = \lambda(x_{p} - x_{r}) - y_{p}\end{gather*}
+        $$
+
+        and $P + Q = R = (x_{r},\ y_{r})$.
```

Comment:
Another indention that I don't believe we support:
![image](https://github.com/matter-labs/zksync-docs/assets/1890113/1306a277-27d8-4c62-a017-c69f374fcd9d)

Reply:
Yeah, we can remove the indentations.